python/pyspark/serializers.py

```diff
@@ -184,27 +184,67 @@ def loads(self, obj):
         raise NotImplementedError
 
 
-class ArrowSerializer(FramedSerializer):
+class BatchOrderSerializer(Serializer):
```

Comment:
@BryanCutler, just read https://github.com/apache/spark/pull/21546#issuecomment-400824224. How much performance gain was made by this? Looks pretty complicated..

Reply:
Yeah, the performance gain by sending out of order batches was small, but the reason this was done was to improve memory usage in the driver JVM.  Before this it still had a worst case of buffering the entire dataset in the JVM, but now nothing is buffered and partitions are immediately sent to Python.  I think that's a huge improvement that is worth the additional complexity. This method might even be applicable to a `collect()` in Python also.

Reply:
Thanks for elaborating this, @BryanCutler. Would you mind if I ask to add this separately in a separate PR? I am actually not super sure on this ..

Reply:
Yeah, I could separate this but is there anything I can do to alleviate your concern?  I'm not sure I'll have the time to try to make another PR before 2.4.0 code freeze and I think this is a really useful memory optimization to help prevent OOM in the driver JVM.  Also, I might have to rerun the benchmarks here, just to be thorough, because the previous ones were from quite a while ago.

Reply:
Ah, okay. I think I understood the benefit. But my impression is that this is something we already were doing. Also, if this is something we could apply to other functionalities too, then it sounded to me a bit of orthogonal work to do separately.

Another concern is, for example, how much we'd likely hit this OOM because I usually expect the data for createDataFrame from Pandas DataFrame or toPandas is likely be small.

If the changes were small, then it would have been okay to me but kind of large changes and looks affecting many codes from Scala side to Python side.

Reply:
Since you verifies the performance difference is trivial, I don't think it's a hard requirement to merge this though. At least, I would just push this in.

Reply:
@HyukjinKwon those are fair questions, not sure I'm going to have the time to do anything about it right now, but I'll circle back to this later.

Reply:
Thank you @BryanCutler.